### PR TITLE
[Yaml] parse merge keys with PARSE_OBJECT_FOR_MAP flag

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24133
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by ostrolucky** [comment]
You use same kind of foreach loops I removed via https://github.com/symfony/symfony/pull/21756 Please consider using union operator again. It reduces big chunk of unnecessary verbosity and you get NULL handling for free. Example of usage in your patch:

```diff
@@ -257,14 +257,8 @@
 
                         $refValue = $this->refs[$refName];
 
-                        if (is_array($refValue)) {
-                            $data += $refValue; // array union
-                        } elseif ((bool) (Yaml::PARSE_OBJECT_FOR_MAP & $flags) && $refValue instanceof \stdClass) {
-                            foreach ($refValue as $refValueKey => $refValueValue) {
-                                if (!isset($data[$refValueKey])) {
-                                    $data[$refValueKey] = $refValueValue;
-                                }
-                            }
+                        if (is_array($refValue) || (bool) (Yaml::PARSE_OBJECT_FOR_MAP & $flags) && $refValue instanceof \stdClass) {
+                            $data += (array) $refValue; // array union
                         } else {
                             throw new ParseException('YAML merge keys used with a scalar value instead of an array.', $this->getRealCurrentLineNb() + 1, $this->currentLine);
                         }
@@ -294,11 +288,7 @@
                                 $data += $parsed; // array union
                             }
                         } elseif ((bool) (Yaml::PARSE_OBJECT_FOR_MAP & $flags) && $parsed instanceof \stdClass) {
-                            foreach ($parsed as $parsedKey => $parsedValue) {
-                                if (!isset($data[$parsedKey])) {
-                                    $data[$parsedKey] = $parsedValue;
-                                }
-                            }
+                            $data += (array) $parsed;
                         } else {
                             throw new ParseException('YAML merge keys used with a scalar value instead of an array.', $this->getRealCurrentLineNb() + 1, $this->currentLine);
                         }
```

**Comment by xabbuh** [comment]
ping @symfony/deciders :)

**Comment by chalasr** [comment]
@xabbuh Just tried it out, it seems to fix it for #24133 code but not for #23587.
Reproducer:

```php
<?php
require __DIR__ .'/vendor/autoload.php';

use Symfony\Component\Yaml\Parser;
use Symfony\Component\Yaml\Yaml;

$yaml = <<<YAML
root:
    mergekeyrefdef:
        a: foo
        <<: &quux
            b: bar
            c: baz
    mergekeyderef:
        d: quux
        <<: *quux
YAML;

$arr = (new Parser())->parse($yaml, Yaml::PARSE_OBJECT_FOR_MAP);
```

**Comment by xabbuh** [comment]
@chalasr Thanks for checking. You are absolutely right. And the bug described in #23587 is actually not related to the `PARSE_OBJECT_FOR_MAP` flag, but purely to the fact that the parser fails when a reference is defined on a merge key. We need to fix that in a different PR on the `2.7` branch.

File: src/Symfony/Component/Yaml/Parser.php [hunk 1]
```diff
@@ -249,14 +249,18 @@ private function doParse($value, $flags)
                 if ('<<' === $key) {
                     $mergeNode = true;
                     $allowOverwrite = true;
-                    if (isset($values['value']) && 0 === strpos($values['value'], '*')) {
+                    if (isset($values['value']) && isset($values['value'][0]) && '*' === $values['value'][0]) {
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: the first `isset` can be removed, as it is covered by the second one.

File: src/Symfony/Component/Yaml/Parser.php [hunk 2]
```diff
@@ -249,14 +249,18 @@ private function doParse($value, $flags)
...
                         $refName = substr(rtrim($values['value']), 1);
                         if (!array_key_exists($refName, $this->refs)) {
                             throw new ParseException(sprintf('Reference "%s" does not exist.', $refName), $this->getRealCurrentLineNb() + 1, $this->currentLine);
                         }
 
                         $refValue = $this->refs[$refName];
 
+                        if ((bool) (Yaml::PARSE_OBJECT_FOR_MAP & $flags) && $refValue instanceof \stdClass) {
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: the explicit `bool` cast is useless (`&&` will trigger it too)

**Review by chalasr — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**